designs/statefulset-disruption.md

```diff
@@ -0,0 +1,891 @@
+
+If Karpenter calls EC2 TerminateInstance **before** EC2 DetachVolume calls from EBS CSI Driver Controller pod finish, then the volumes won't be detached **until the old instance terminates**.This delay depends on how long it takes the underlying instance to enter the `terminated` state, which depends on the instance type. Typically 1 minute for `m5a.large`, up to 15 minutes for certain Metals/GPU/Windows instances. See [appendix D1](#d1-ec2-termination--ec2-detachvolume-relationship-) for more context. 
+ 
+Customers will see the following events (Note the 1-min delay between `Multi-Attach error` AND `AttachVolume.Attach failed`):
+
+```
+Warning  FailedAttachVolume      102s               attachdetach-controller  Multi-Attach error...
+Warning  FailedAttachVolume      40s                attachdetach-controller  AttachVolume.Attach failed for volume "pvc-" : rpc error: code = Internal desc = Could not attach volume "vol-" to node "i-"... VolumeInUse: vol- is already attached to an instance                    
+Normal   SuccessfulAttachVolume  33s                attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc"                                                                   
+```
+
+Customers can determine which delay they are suffering from based off of whether `AttachVolume.Attach` is in the `FailedAttachVolume` event. 
+
+### Solutions
+
+**A1: To solve A long-term, Kubernetes should ensure volumes are unmounted before critical pods like CSI Driver Node pod are terminated.**
+
+**A2: To solve A today, Karpenter should confirm that volumes are not in use and confirm AttachDetach Controller knows this before deleting the node's finalizer.**
+
+**B1: To solve B today, Karpenter should wait for volumes to detach before terminating the node.**
+
+See [WIP Kubernetes 1.31 A1 solution in PR #125070](https://github.com/kubernetes/kubernetes/pull/125070)
+
+See [a proof-of-concept implementation of A2 & B1 in PR #1294](https://github.com/kubernetes-sigs/karpenter/pull/1294)
+
+Finally, we should add the following EBS x Karpenter end-to-end test in karpenter-provider-aws to catch regressions between releases of Karpenter or EBS CSI Driver:
+1. Deploy statefulset with 1 replica
+2. Consolidate Node
+3. Confirm replica migrated
+4. Confirm replica running within x minutes (where x is high enough to prevent flakes)
+
+## Problem A. Preventing 6+ minute delays 
+
+If `ReadWriteOnce` volumes are not unmounted *by CSI Node Service*, Kubernetes cannot confirm volumes are not in use and safe to attach to new node. Kubernetes will wait 6 minutes[^4] and ensure Node is `unhealthy` before treating the volume as unmounted and moving forward with a volume detach. See [EBS CSI 6-minute delay FAQ](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/docs/faq.md#6-minute-delays-in-attaching-volumes) 
+
+Cluster operator will see a `FailedAttachVolume` event with `Multi-Attach error`
```

Comment:
Is this a kube-event?

Reply:
This is an event on PVC and Pod object if that answers your question.